vSphere DRS Anti-Affinity Rule Configuration During UPI installation (#51584)
openshift/openshift-docs: modules/vsphere-anti-affinity.adoc

// Module included in the following assemblies:
//
// * storage/persistent_storage/persistent-storage-vsphere.adoc
// * installing/installing_vmc/installing-restricted-networks-vmc-user-infra.adoc
// * installing/installing_vmc/installing-vmc-customizations.adoc
// * installing/installing_vmc/installing-vmc-network-customizations.adoc
// * installing/installing_vmc/installing-vmc.adoc
// * installing/installing_vmc/installing-vmc-network-customizations-user-infra.adoc
// * installing/installing_vmc/installing-vmc-user-infra.adoc
// * installing/installing_vsphere/installing-restricted-networks-vsphere.adoc
// * installing/installing_vsphere/installing-vsphere-installer-provisioned-customizations.adoc
// * installing/installing_vsphere/installing-vsphere-installer-provisioned-network-customizations.adoc
// * installing/installing_vsphere/installing-vsphere-installer-provisioned.adoc
// * installing/installing_vsphere/installing-vsphere-network-customizations.adoc
// * installing/installing_vsphere/installing-vsphere.adoc

:_content-type: PROCEDURE
[id="anti-affinity-vsphere_{context}"]
= Configure vSphere DRS anti-affinity rules for control plane nodes

> {product-title} Distributed Resource Scheduler (DRS) anti-affinity rules can be configured to support higher availability of OpenShift Control Plane nodes so that the vSphere Virtual Machines for these nodes are not scheduled to the same vSphere Host.

Reviewer: Can we cross-link to the doc below for reference? 

https://github.com/openshift/openshift-docs/pull/51056

Author: @bostrt I can't install a cross-link in a module (this keeps modules portable, which this one will be once I move it to UPI section from IPI section). However there are other ways we can refer to other places in the documentation. I'll look to see what I'm able to do.

Author: (and BTW, @bostrt I know that seems...oddly primitive in this day and age. It's because of the post-processing for the customer portal. I dislike it myself.)

Author: Hey @bostrt the doc produced by that PR already appears "above" and as a prerequisite in all the installation sections, including the one that this section is going into. Because of that, may I avoid mentioning it in our PR?

Author: Thank you sir....

.Example command

[source,terminal]
----
$ govc cluster.rule.create \
  -name openshift4-control-plane-group \
  -dc MyDatacenter -cluster MyCluster \
  -enable \
  -anti-affinity master-0 master-1 master-2

> ----

Reviewer: @GroceryBoyJr here's example successful output from this command:
```
[13-10-22 09:14:41] Reconfigure /MyDatacenter/host/MyCluster...OK
```

Author: Thank you, @bostrt . Update completed!

After creating the rule, your control plane nodes will be automatically migrated by vSphere so they are not running on the same hosts. This may take some time while vSphere reconciles the new rule.

[NOTE]
====
The migration will occur automatically and may cause brief OpenShift API outage or latency until the migration is completed by vSphere.
====

== Maintaining vSphere DRS anti-affinity rules

The vSphere DRS Anti-Affinity rules will need to be updated manually in the event of a control plane VM name change or migration to new a vSphere Cluster.

.Procedure

. Remove existing rule:
+
[source,terminal]
----
$ govc cluster.rule.remove \
  -name my-control-plane-group \
  -dc MyDatacenter -cluster MyCluster

> ----

Reviewer: @GroceryBoyJr example successful output looks the same when removing:
```
[13-10-22 09:14:41] Reconfigure /MyDatacenter/host/MyCluster...OK
```

Author: @bostrt I've added this example output to the page.

// Module included in the following assemblies:
//
// * installing/installing_vsphere/installing-restricted-networks-vsphere.adoc
// * installing/installing_vsphere/installing-vsphere-network-customizations.adoc
// * installing/installing_vsphere/installing-vsphere.adoc

:_content-type: PROCEDURE
[id="anti-affinity-vsphere_{context}"]
= Configuring vSphere DRS anti-affinity rules for control plane nodes

vSphere Distributed Resource Scheduler (DRS) anti-affinity rules can be configured to support higher availability of OpenShift Control Plane nodes so that the vSphere Virtual Machines for these nodes are not scheduled to the same vSphere Host. 

[IMPORTANT]
====
* The following information applies to compute DRS only and does not apply to storage DRS.

* The `govc` command is an open-source command available from VMware; it is not available from Red Hat. `govc` is not supported by the Red Hat support. 

> * Instructions for downloading and installing 'govc' are found on the VMware documentation website.

Reviewer 2: govc is in quotes here, but should be in backticks to match the rest

==== 

// https://docs.vmware.com/en/VMware-Telco-Cloud-Operations/1.4.0/deployment-guide-140/GUID-5249E662-D792-4A1A-93E6-CF331552364C.html#:~:text=Govc%20is%20an%20open%20source,operations%20on%20the%20target%20vCenter.

.Example command

[source,terminal]
----
# govc cluster.rule.create \
  -name openshift4-control-plane-group \
  -dc MyDatacenter -cluster MyCluster \
  -enable \
  -anti-affinity master-0 master-1 master-2
----

> After creating the rule, your control plane nodes will be automatically migrated by vSphere so they are not running on the same hosts. This might take some time while vSphere reconciles the new rule. Successful command completion is shown below:

Reviewer 2: IBMSG recommends against using directions (i.e., "shown below). This mainly has to do with screen readers/non-traditional readers. So maybe we could reword it: 

"Successful command completion is shown in the following procedure." 

I also don't think this would get a colon, since it doesn't go directly into the procedure.

Author: Thank you! Done!

[NOTE]
====
The migration will occur automatically and might cause brief OpenShift API outage or latency until the migration is completed by vSphere.
====

The vSphere DRS anti-affinity rules will need to be updated manually in the event of a control plane VM name change or migration to a new vSphere Cluster.

.Procedure

> . Remove any existing DRS anti-affinity rule:

Reviewer 2: is `govc` a command? If it is, could we say: 

". Remove any existing DRS anti-affinity rule by entering the following command:" ? This goes along with OCP guidelines: https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#formatting

According to Google it is. ALL of your commands need updated with "by entering the following command: or "by running the following command:"

Author: Yes, govc is a command. Thank you, this has been completed.

// Module included in the following assemblies:
//
// * installing/installing_vsphere/installing-restricted-networks-vsphere.adoc
// * installing/installing_vsphere/installing-vsphere-network-customizations.adoc
// * installing/installing_vsphere/installing-vsphere.adoc

:_content-type: PROCEDURE
[id="anti-affinity-vsphere_{context}"]
= Configuring vSphere DRS anti-affinity rules for control plane nodes

> vSphere Distributed Resource Scheduler (DRS) anti-affinity rules can be configured to support higher availability of OpenShift Control Plane nodes so that the vSphere Virtual Machines for these nodes are not scheduled to the same vSphere Host. 

Reviewer 3: This is a really long sentence, and I am not sure if the anti-affinity rules ensure the nodes are scheduled correctly or if that is a benefit of high availability. I think it is the former, but I can see this being an issue in translation. WDYT of the following? Also, should _OpenShift_ be  `{product-title}`?

```suggestion
vSphere Distributed Resource Scheduler (DRS) anti-affinity rules can be configured to support higher availability of {product-title} Control Plane nodes. Anti-affinity rules ensure that the vSphere Virtual Machines for the {product-title} Control Plane nodes are not scheduled to the same vSphere Host. 
```

Author: Done, Thank You!

[IMPORTANT]
====
* The following information applies to compute DRS only and does not apply to storage DRS.

> * The `govc`` command is an open-source command available from VMware; it is not available from Red Hat. `govc` is not supported by the Red Hat support. 

Reviewer 3: ```suggestion
* The `govc`` command is an open-source command available from VMware; it is not available from Red Hat. The `govc` command is not supported by Red Hat support. 
```

Author: Done, Thank You!

* Instructions for downloading and installing `govc` are found on the VMware documentation website.
==== 

// https://docs.vmware.com/en/VMware-Telco-Cloud-Operations/1.4.0/deployment-guide-140/GUID-5249E662-D792-4A1A-93E6-CF331552364C.html#:~:text=Govc%20is%20an%20open%20source,operations%20on%20the%20target%20vCenter.

Create an anti-affinity rule by running the following command:

.Example command

[source,terminal]
----
$ govc cluster.rule.create \
  -name openshift4-control-plane-group \
  -dc MyDatacenter -cluster MyCluster \
  -enable \
  -anti-affinity master-0 master-1 master-2
----

> After creating the rule, your control plane nodes will be automatically migrated by vSphere so they are not running on the same hosts. This might take some time while vSphere reconciles the new rule. Successful command completion is shown in the following procedure.

Reviewer 3: ```suggestion
After creating the rule, your control plane nodes are automatically migrated by vSphere so they are not running on the same hosts. This might take some time while vSphere reconciles the new rule. Successful command completion is shown in the following procedure.
```

Reviewer 3: See ISG guidance on [future tense](https://www.ibm.com/docs/en/ibm-style?topic=grammar-verbs#tense).

Author: Done, thank you!

[NOTE]
====

> The migration will occur automatically and might cause brief OpenShift API outage or latency until the migration is completed by vSphere.

Reviewer 3: ```suggestion
The migration occurs automatically and might cause brief OpenShift API outage or latency until the migration finishes.
```
See[ ISG guidance](https://www.ibm.com/docs/en/ibm-style?topic=word-usage#complete) on _complete_.

Author: Done, Thank You!

====

> The vSphere DRS anti-affinity rules will need to be updated manually in the event of a control plane VM name change or migration to a new vSphere Cluster.

Author: Done, Thank You!

// Module included in the following assemblies:
//
// * installing/installing_vsphere/installing-restricted-networks-vsphere.adoc
// * installing/installing_vsphere/installing-vsphere-network-customizations.adoc
// * installing/installing_vsphere/installing-vsphere.adoc

:_content-type: PROCEDURE
[id="anti-affinity-vsphere_{context}"]

> = Configure vSphere DRS anti-affinity rules for control plane nodes

Reviewer 4: If you do keep this as a procedure, it should start with a gerund: "Configuring"

Author: Done, Thank You!

vSphere Distributed Resource Scheduler (DRS) anti-affinity rules can be configured to support higher availability of OpenShift Control Plane nodes so that the vSphere Virtual Machines for these nodes are not scheduled to the same vSphere Host. 

[IMPORTANT]
====
* The following information applies to compute DRS only and does not apply to storage DRS.

> * The `govc` command is an open-source command available from VMware; it is not available from Red Hat. `govc` is not supported by the Red Hat Enterprise Linux technical support organization. 

Reviewer 4: I think we just say "by Red Hat Support".

Author: Oh that's way better, thank you.

* Instructions for downloading and installing 'govc' are found on the VMware documentation website.
==== 

// https://docs.vmware.com/en/VMware-Telco-Cloud-Operations/1.4.0/deployment-guide-140/GUID-5249E662-D792-4A1A-93E6-CF331552364C.html#:~:text=Govc%20is%20an%20open%20source,operations%20on%20the%20target%20vCenter.

.Example command

> [source,terminal]

Reviewer 4: Is there a reason that you're giving the command and output both here and in the procedure below?

I'd recommend removing them from up here and leaving those details to the procedure

Author: Good point, done. Thank you!

----
# govc cluster.rule.create \
  -name openshift4-control-plane-group \
  -dc MyDatacenter -cluster MyCluster \
  -enable \
  -anti-affinity master-0 master-1 master-2
----

After creating the rule, your control plane nodes will be automatically migrated by vSphere so they are not running on the same hosts. This might take some time while vSphere reconciles the new rule. Successful command completion is shown below:

.Example output
[source,terminal]
----
 [13-10-22 09:14:41] Reconfigure /MyDatacenter/host/MyCluster...OK
----

[NOTE]
====
The migration will occur automatically and might cause brief OpenShift API outage or latency until the migration is completed by vSphere.
====

> == Maintaining vSphere DRS anti-affinity rules

Reviewer 4: Procedure modules shouldn't have a subheading in them like this.

It seems like what you have above this could probably be broken out into a separate module. I won't hold up the PR over it being all in one file like this, but it's something you should consider taking a look of in the future.

But the subheading here should definitely be removed before merging. If you really do want that subheading, then that would be a good indication that you might want to break them up into two separate modules

Author: I have removed this subheading. Thank you!

> The vSphere DRS Anti-Affinity rules will need to be updated manually in the event of a control plane VM name change or migration to a new vSphere Cluster.

Reviewer 4: lowercase anti-affinity to match rest of the module

Author: Done, Thank You!

.Procedure

. Remove any existing DRS anti-affinity rule:
+
[source,terminal]
----

> # govc cluster.rule.remove \

Reviewer 4: Does this really have to be run as root? If not, it should be "$". If yes, then you should explicit say that in a prepreq or a step. See the SSG on this topic: https://redhat-documentation.github.io/supplementary-style-guide/#commands-with-root-privileges

(several places throughout the PR, if an update is required)

Author: @bergerhoffer Unfortunately, yes, it has to be run as root. It's a vmware utility that influences control at the hypervisor level.  https://docs.vmware.com/en/VMware-Telco-Cloud-Operations/1.4.0/deployment-guide-140/GUID-5249E662-D792-4A1A-93E6-CF331552364C.html  Yes it gives me indigestion but QE confirmed the requirement also. :-/

Reviewer 4: @GroceryBoyJr That's totally fine if it does need to be run as root! The guidance I linked is just saying that instead of only relying on the `#` prompt to indicate to run as a root user, you should also explicitly say that it is required in the step or in the prereqs.

I see that you change the `#` to `$`, and I don't see any text saying that root is required. Does that prompt need to be changed back, if root is really required?

(If you do have to change it back and add a sentence about being root, there is a double backtick in your PR now that could also be fixed)

  -name openshift4-control-plane-group \
  -dc MyDatacenter -cluster MyCluster
----
+

> The example output appears identical to successful prior application of the rule:

Reviewer 4: I re-read this a few times and don't quite understand what the wording here means.

Are you trying to say that the output for the user should be similar to the example output below? If so, I'd consider rewording to be clearer

Author: I removed the line entirely, it adds nothing. Thank you!

// Module included in the following assemblies:
//
// * installing/installing_vsphere/installing-restricted-networks-vsphere.adoc
// * installing/installing_vsphere/installing-vsphere-network-customizations.adoc
// * installing/installing_vsphere/installing-vsphere.adoc

:_content-type: PROCEDURE
[id="anti-affinity-vsphere_{context}"]
= Configure vSphere DRS anti-affinity rules for control plane nodes

vSphere Distributed Resource Scheduler (DRS) anti-affinity rules can be configured to support higher availability of OpenShift Control Plane nodes so that the vSphere Virtual Machines for these nodes are not scheduled to the same vSphere Host. 

[IMPORTANT]
====
The following information applies to compute DRS only and does not apply to storage DRS.
====

[NOTE]
====

> The `govc` command is an open-source command available from VMware; it is not available from Red Hat. `govc` is not supported by the Red Hat Enterprise Linux technical support organization. 

Reviewer 2: You COULD put these both in the IMPORTANT admonition, and separate them with bullet points (*). That would avoid stacked admonitions. I don't have a preference either way.

Author: Done, thank you!

Instructions for downloading and installing 'govc' are found on the VMware documentation website.
==== 

// https://docs.vmware.com/en/VMware-Telco-Cloud-Operations/1.4.0/deployment-guide-140/GUID-5249E662-D792-4A1A-93E6-CF331552364C.html#:~:text=Govc%20is%20an%20open%20source,operations%20on%20the%20target%20vCenter.

.Example command

[source,terminal]
----
$ govc cluster.rule.create \
  -name openshift4-control-plane-group \
  -dc MyDatacenter -cluster MyCluster \
  -enable \
  -anti-affinity master-0 master-1 master-2
----

> After creating the rule, your control plane nodes will be automatically migrated by vSphere so they are not running on the same hosts. This may take some time while vSphere reconciles the new rule. Successful command completion is shown below:

Reviewer 2: IBMSG recommends avoiding "may"

Author: Changed to "might". Thank you!

.Example output
[source,terminal]
----

> $ [13-10-22 09:14:41] Reconfigure /MyDatacenter/host/MyCluster...OK

Reviewer 2: Would the "$" be included in the example output?

Author: No, it should be # as suggested by QE. Changed them all. Thank You!

Reviewer 5: @GroceryBoyJr Pls just remove the "$" sign.

Author: OK! Got it!

----

[NOTE]
====

> The migration will occur automatically and may cause brief OpenShift API outage or latency until the migration is completed by vSphere.

Reviewer 2: Another instance of "may". Swap for "might"

Author: Done! Thank you!

====

== Maintaining vSphere DRS anti-affinity rules

The vSphere DRS Anti-Affinity rules will need to be updated manually in the event of a control plane VM name change or migration to new a vSphere Cluster.

.Procedure

> . Remove existing rule:

Reviewer 2: the existing?

Author: @stevsmit Great point. I changed it even more to clarify: "Remove any existing DRS anti-affinity rule:"

+
[source,terminal]
----
$ govc cluster.rule.remove \
  -name openshift4-control-plane-group \
  -dc MyDatacenter -cluster MyCluster
----
+
The example output appears identical to successful application of the rule:
+
[source,terminal]
----

> $ [13-10-22 09:33:24] Reconfigure /MyDatacenter/host/MyCluster...OK

Reviewer 2: Same thing here. Would the "$" be present?

Reviewer 5: "$" should not present, here's the output sample:
```
# govc cluster.rule.remove -name openshift4-control-plane-group -dc OCP-DC -cluster workloads
[25-10-22 05:03:19] Reconfigure /OCP-DC/host/qe-cluster/workloads...OK
```

Author: Done! Thank you both, @stevsmit and @sgaoshang

Reviewer 5: @GroceryBoyJr Maybe the sample I provided made a confusion, pls don't use it directly. The origin output is much more appropriate, just remove the "$" sign.

Reviewer 5: Change `$ [13-10-22 09:33:24] Reconfigure /MyDatacenter/host/MyCluster...OK`
to 
`[13-10-22 09:33:24] Reconfigure /MyDatacenter/host/MyCluster...OK`

Author: Got it, thank  you, this is done!

----

> . Create rule again with updated names:

Reviewer 2: the rule?

Author: @stevsmit Thank you, done.

.Example command

[source,terminal]
----
$ govc cluster.rule.create \
  -name openshift4-control-plane-group \
  -dc MyDatacenter -cluster MyCluster \
  -enable \
  -anti-affinity master-0 master-1 master-2
----

After creating the rule, your control plane nodes will be automatically migrated by vSphere so they are not running on the same hosts. This may take some time while vSphere reconciles the new rule. Successful command completion is shown below:

.Example output
[source,terminal]
----
$ [13-10-22 09:14:41] Reconfigure /MyDatacenter/host/MyCluster...OK
----

[NOTE]
====
The migration will occur automatically and may cause brief OpenShift API outage or latency until the migration is completed by vSphere.
====

== Maintaining vSphere DRS anti-affinity rules

The vSphere DRS Anti-Affinity rules will need to be updated manually in the event of a control plane VM name change or migration to new a vSphere Cluster.

.Procedure

. Remove existing rule:
+
[source,terminal]
----
$ govc cluster.rule.remove \

>   -name my-control-plane-group \

Reviewer 5: It's better to use consistent rule name in example, I see here is `my-control-plane-group` while others are `openshift4-control-plane-group`.

Author: Thank you for that feedback, @sgaoshang ! Much appreciated!
